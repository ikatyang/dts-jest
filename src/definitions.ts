import * as _ts from 'typescript';

// tslint:disable-next-line:no-var-requires
export const package_name = require('../package.json').name;

export const config_namespace = '_dts_jest_';
export const runtime_namespace = '_dts_jest_runtime_';

export const env_root_dir = 'DTS_JEST_ROOT_DIR';

export const runtime_indent_spaces = 2;

export const trigger_regex = /^\s*\/\/\s*@dts-jest\b(:?\S*)\s*(.+)?\s*$/;
export enum TriggerMatchIndex {
  Input,
  Flags,
  Description,
}
export type TriggerMatchArray = [string, string, string | undefined];

export enum AssertionFlag {
  Snapshot = ':snapshot',
  Show = ':show',
  Pass = ':pass',
  Fail = ':fail',
}
export enum ActualAssertionFlag {
  Error = ':error',
  NoError = ':no-error',
  Question = '?',
}

export enum TestFlag {
  Test = ':test',
  Only = ':only',
  Skip = ':skip',
}
export enum TestMethod {
  Test = 'test',
  Only = 'test.only',
  Skip = 'test.skip',
}
export const group_flag = ':group';
export enum GroupMethod {
  Test = 'describe',
  Only = 'describe.only',
  Skip = 'describe.skip',
}

export interface Target {
  /**
   * 0-based
   */
  line: number;
  expression: string;
  description?: string;
  group?: Group;
}

export interface Snapshot {
  inference?: string;
  diagnostic?: string;
}
export interface Expected {
  value: string;
}

export interface TriggerInfo {
  flags: AssertionFlag[];
  method: TestMethod;
}
export interface Trigger extends Target, TriggerInfo {
  start: number;
  end: number;
}
export interface ActualTrigger extends Trigger, Expected {}

export interface GroupInfo {
  method: GroupMethod;
}
export interface Group extends GroupInfo {
  title: string;
}

export type TriggerOrGroupInfo =
  | ({ is_group: false } & TriggerInfo)
  | ({ is_group: true } & GroupInfo);

export interface Result extends Target, Snapshot {}
export interface ActualResult extends Target, Expected {}

export interface JestConfig {
  rootDir: string;
  globals: { [K in typeof config_namespace]?: RawConfig };
  _dts_jest_internal_test_?: boolean;
}

export interface RawConfig {
  tsconfig?: string | Record<string, _ts.CompilerOptionsValue>;
  type_format?: _ts.TypeFormatFlags;
  typescript?: string;
}

export interface FormattedConfig {
  tsconfig: _ts.CompilerOptions;
  type_format: _ts.TypeFormatFlags;
  typescript: typeof _ts;
  typescript_path: string;
}

export type JestSnapshot = Record<string, string>;
