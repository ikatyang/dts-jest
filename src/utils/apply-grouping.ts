import { TriggerGroup } from '../definitions';
import { create_group_expression } from './create-group-expression';

export const apply_grouping = (
  line_contents: string[],
  groups: TriggerGroup[],
) => {
  const grouped_line_contents = line_contents.slice();

  if (groups.length === 0) {
    return grouped_line_contents;
  }

  interface GroupScope {
    start_line: number;
    end_line: number;
    group: TriggerGroup;
  }

  const scopes: GroupScope[] = [];

  if (groups.length === 1) {
    const [group] = groups;
    scopes.push({
      group,
      start_line: group.line,
      end_line: line_contents.length - 1,
    });
  } else {
    let last_group = groups[0];

    for (let i = 1; i < groups.length; i++) {
      const group = groups[i];

      scopes.push({
        group: last_group,
        start_line: last_group.line,
        end_line: group.line,
      });

      last_group = group;
    }

    scopes.push({
      group: last_group,
      start_line: last_group.line,
      end_line: line_contents.length - 1,
    });
  }

  scopes
    .map(scope => {
      const [group_expression_left, group_expression_right] =
        `${create_group_expression(scope.group, '\n')};`.split('\n');

      grouped_line_contents[scope.start_line] =
        group_expression_left + grouped_line_contents[scope.start_line];

      return { scope, group_expression_right };
    })
    .forEach(({ scope, group_expression_right }, index) => {
      grouped_line_contents[scope.end_line] =
        index === scopes.length - 1
          ? grouped_line_contents[scope.end_line] + group_expression_right
          : group_expression_right + grouped_line_contents[scope.end_line];
    });

  return grouped_line_contents;
};
