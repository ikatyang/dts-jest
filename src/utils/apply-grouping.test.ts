import { GroupMethod } from '../definitions';
import { apply_grouping } from './apply-grouping';

it('should return correctly with groups (length = 0)', () => {
  expect(
    apply_grouping(['/* line 1 */', '/* line 2 */', '/* line 3 */'], []),
  ).toMatchSnapshot();
});

it('should return correctly with groups (length = 1)', () => {
  expect(
    apply_grouping(
      ['/* line 1 */', '/* line 2 */', '/* line 3 */'],
      [{ line: 1, method: GroupMethod.Test, description: 'group L2' }],
    ),
  ).toMatchSnapshot();
});

it('should return correctly with groups (length > 1)', () => {
  expect(
    apply_grouping(
      [
        '/* line 1 */',
        '/* line 2 */',
        '/* line 3 */',
        '/* line 4 */',
        '/* line 5 */',
        '/* line 6 */',
        '/* line 7 */',
        '/* line 8 */',
        '/* line 9 */',
      ],
      [
        { line: 2, method: GroupMethod.Test, description: 'group L3' },
        { line: 4, method: GroupMethod.Test, description: 'group L5' },
        { line: 7, method: GroupMethod.Test, description: 'group L8' },
      ],
    ),
  ).toMatchSnapshot();
});
