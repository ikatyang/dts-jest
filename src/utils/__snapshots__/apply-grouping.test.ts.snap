// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly with groups (length = 0) 1`] = `
Array [
  "/* line 1 */",
  "/* line 2 */",
  "/* line 3 */",
]
`;

exports[`should return correctly with groups (length = 1) 1`] = `
Array [
  "/* line 1 */",
  "describe(\\"group L2\\", function () { /* line 2 */",
  "/* line 3 */ });",
]
`;

exports[`should return correctly with groups (length > 1) 1`] = `
Array [
  "/* line 1 */",
  "/* line 2 */",
  "describe(\\"group L3\\", function () { /* line 3 */",
  "/* line 4 */",
  " });describe(\\"group L5\\", function () { /* line 5 */",
  "/* line 6 */",
  "/* line 7 */",
  " });describe(\\"group L8\\", function () { /* line 8 */",
  "/* line 9 */ });",
]
`;
